server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -212,6 +213,7 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
     private final IndexEventListener indexEventListener;
     private final QueryCachingPolicy cachingPolicy;
     private final Supplier<Sort> indexSortSupplier;
+    private final Set<String> fieldHasValue;
```

Comment:
We could move this Set into IndicesService in order to keep one Set per index instead of one Set per shard as this current implementation.
The advantage of the current impl is that even if a shard is relocated to a different data node the response is correct and persistent for all the data nodes. 
The advantage of moving this Set at the index level is that we are not duplicating information and therefore using less heap. In the second case, even if a shard is relocated to a different data node and therefore the response for the current data node could be wrong (if the only document with one field is relocated) the overall field caps response is correct.

Reply:
Keeping track of the info at the indices service level, hence for all shards of a certain index that are allocated on a certain data node: that is accurate despite relocations because we track non empty field names, and a field never becomes empty after being non empty (we don't account for deletions), hence even if a single shard that gets relocated was contributing to a certain field being non empty, the end result would still be at the index level that that field is non empty. Correct?

Reply:
That's correct! After talking to @original-brownbear we agreed to save such info at indices service level.

Reply:
I am fine with this: let's document in the code the choice we made and the constraints / limitations, just do we don't end up thinking we got it wrong one day in the future.

Reply:
Also I am curious to see what this means in practice, code-wise. Will the field caps fetcher need updating? How will it get this information for a certain shard given the info is tracked at the indices service level?

Reply:
> let's document in the code the choice we made and the constraints / limitations

Documented!
>Also I am curious to see what this means in practice, code-wise. Will the field caps fetcher need updating? How will it get this information for a certain shard given the info is tracked at the indices service level?

So code-wise instead of declaring a Set for each shard I do declare it at IndexService level and pass it down to each shard when we instantiate IndexShard. This makes it really easy to test and no changes are needed to the field caps fetcher code between the two implementations.